src/native/s3_client.c

```diff
@@ -0,0 +1,394 @@
+
+        if (aws_jni_check_and_clear_exception(env)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_S3_META_REQUEST,
+                "id=%p: Ignored Exception from S3Client.onShutdownCompete callback",
+                (void *)callback->java_s3_client);
+        }
+    }
+
+    // We're done with this callback data, free it.
+    (*env)->DeleteGlobalRef(env, callback->java_s3_client);
+    aws_mem_release(aws_jni_get_allocator(), user_data);
+}
+
+static int s_on_s3_meta_request_body_callback(
+    struct aws_s3_meta_request *meta_request,
+    const struct aws_byte_cursor *body,
+    uint64_t range_start,
+    void *user_data) {
+    (void)body;
+    (void)range_start;
+    int return_value = AWS_OP_ERR;
+
+    uint64_t range_end = range_start + body->len;
+
+    struct s3_client_make_meta_request_callback_data *callback_data =
+        (struct s3_client_make_meta_request_callback_data *)user_data;
+
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+
+    jobject jni_payload = aws_jni_direct_byte_buffer_from_raw_ptr(env, body->ptr, body->len);
+    jint body_response_result = 0;
+
+    if (callback_data->java_s3_meta_request_response_handler_native_adapter != NULL) {
+        body_response_result = (*env)->CallIntMethod(
+            env,
+            callback_data->java_s3_meta_request_response_handler_native_adapter,
+            s3_meta_request_response_handler_native_adapter_properties.onResponseBody,
+            jni_payload,
+            range_start,
+            range_end);
+        (void)body_response_result;
+
+        if (aws_jni_check_and_clear_exception(env)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_S3_META_REQUEST,
+                "id=%p: Ignored Exception from S3MetaRequest.onResponseBody callback",
+                (void *)meta_request);
+            goto cleanup;
+        }
+    }
+    return_value = AWS_OP_SUCCESS;
+
+cleanup:
+    (*env)->DeleteLocalRef(env, jni_payload);
+
+    return return_value;
+}
+
+static int s_on_s3_meta_request_headers_callback(
+    struct aws_s3_meta_request *meta_request,
+    const struct aws_http_headers *headers,
+    int response_status,
+    void *user_data) {
+    int return_value = AWS_OP_ERR;
+    struct s3_client_make_meta_request_callback_data *callback_data =
+        (struct s3_client_make_meta_request_callback_data *)user_data;
+
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+    struct aws_allocator *allocator = aws_jni_get_allocator();
+    /* calculate initial header capacity */
+    size_t headers_initial_capacity = 0;
+    for (size_t header_index = 0; header_index < aws_http_headers_count(headers); ++header_index) {
+        struct aws_http_header header;
+        aws_http_headers_get_index(headers, header_index, &header);
+        /* aws_marshal_http_headers_to_dynamic_buffer() impl drives this calculation */
+        headers_initial_capacity += header.name.len + header.value.len + 8;
```

Comment:
what's the 8?  maybe make a constant with a nice name

Reply:
The aforementioned function's logic writes both the lengths of the header name(4 bytes) and header value(4 bytes) so the 8 is that sum on top of the name and value data. I'm not sure it's sensible to make a constant given the code comment above

Reply:
I'd still add something one way or another; it's better if someone doesn't need to cross-reference across files just to feel confident.